### ISSUE TITLE
Enhanced bucket_counts script

### ISSUE THREAD

**Issue by amateur-professional** [crits/crits#859]
Hey dev team,

Our instance of CRITs makes extensive use of the bucket list system - as such it was exceedingly necessary to have accurate counts of each bucket. However, the stock bucket_counts script is exceedingly memory-hungry - I once saw it eat more than 20 GB on our dataset, and we don't exactly have a big instance. Growing frustrated with our memory issues I took it upon myself to refactor the bucket_counts script - what used to take more than an hour and 20GB of RAM now takes roughly 3 seconds and less than 100MB on our dataset. I'd normally submit a pull request for this, but our instance is too customized and I'm not so good with git. The script is as follows: 

```
from django.core.management.base import BaseCommand

from crits.core.bucket import Bucket
from crits.core.class_mapper import class_from_type

class Command(BaseCommand):
    """
    Script Class.
    """
    def handle(self, *args, **options):
        """
        Script Execution.
        """
        # create new dict to hold bucket names and their values
        buckets = {}
        # define object types to count
        types = ['Actor', 'Campaign', 'Certificate', 'Domain', 'Email', 'Event',
                 'Indicator', 'IP', 'PCAP', 'RawData', 'Signature', 'Sample', 'Target']
        # iterate through otypes, building a list of Buckets
        for otype in types:
            klass = class_from_type(otype)
            if not klass:
                continue
            # get list of distinct bucket list names for this otype, only get bucket lists that are not empty, then get unique buckets - this works because distinct() on ListFields
            # works to get unique list entries, not unique lists
            bucket_list = klass.objects().filter(bucket_list__ne='[]').distinct('bucket_list')
            # check to see if list length is greater than zero
            if len(bucket_list) > 0:
                # iterate through buckets found for this otype
                for bucket in bucket_list:
                    # sanity check for buckets that are None
                    if bucket is None:
                        break # we don't want to continue
                    # if this is a new bucket, add it to the list and set the value for this otype
                    if bucket not in buckets:
                        buckets[bucket] = Bucket()
                        buckets[bucket].name = bucket
                        setattr(buckets[bucket], otype, 0) # we know the bucket exists for this otype - but don't know how many yet
                    # if it's not a new bucket, that means we found it for a different otype and need to set the count for this otype
                    else:
                        setattr(buckets[bucket], otype, 0)
            # if list is zero, just continue iterating through otypes
            else:
                continue


        # now we have all the distinct names of buckets for all otypes now we need to count how many of each type there is
        # iterate through buckets
        for bucket in buckets:
            # iterate through otypes
            for otype in types:
                # check if the bucket we are looking at definitely has objects with this otype - we already set this if it has
                if hasattr(buckets[bucket], otype):
                    # map the appropriate class
                    klass = class_from_type(otype)
                    # get a count of objects of this type that have this bucket
                    count = klass.objects(bucket_list=buckets[bucket].name).count()
                    # assign that count to this bucket
                    setattr(buckets[bucket], otype, count)
                # if the bucket does not have objects with this type, set the count for this type to 0
                else:
                    setattr(buckets[bucket], otype, 0)
        # Drop all existing buckets
        Bucket.objects().delete()
        # save all stored buckets
        for bucket in buckets.values():
            bucket.save()
```

**Comment by brlogan** [comment]
Nice work!  I'd be curious to see how your solution compares to something like:

```python
pipe = [{'$match': {'bucket_list': {'$exists': True},
                    'bucket_list': {'$ne': []}}},
        {'$unwind': '$bucket_list'},
        {'$group': {'_id': '$bucket_list', 'count': {'$sum': 1}}}]

# This will only work with Mongoengine 0.9+
buckets = klass.objects().aggregate(pipeline=pipe)

# For Mongoengine <0.9
coll = mongo_connector(klass._meta['collection'])
buckets = coll.aggregate(pipeline=pipe)
```

Obviously you'd still have to walk the list and generate and save the Bucket objects.

**Comment by dsnellgrove** [comment]
I totally forgot to send back a PR for this. We did a change related to this about two years ago. I used an aggregation for performance as well.

```
from django.core.management.base import BaseCommand

from crits.core.bucket import Bucket
from crits.core.class_mapper import class_from_type

class Command(BaseCommand):
    """
    Script Class.
    """

    def handle(self, *args, **options):
        """
        Script Execution.
        """

        buckets = {}

        types = ['Campaign', 'Certificate', 'Domain', 'Email', 'Event',
                 'Indicator', 'IP', 'PCAP', 'RawData', 'Sample', 'Target']

        for otype in types:
            klass = class_from_type(otype)
            if not klass:
                continue
            bucket_counts = klass._get_collection().aggregate([{'$unwind': '$bucket_list'}, {'$group': {'_id': '$bucket_list', 'count': {'$sum': 1}}}])
            for bucket_count in bucket_counts['result']:
                bname = bucket_count['_id']
                if not bname:
                    continue # Avoid empty strings
                if bname not in buckets:
                    buckets[bname] = Bucket()
                    buckets[bname].name = bname
                setattr(buckets[bname], otype, bucket_count['count'])

        # Drop all existing buckets
        Bucket.objects().delete()

        for bucket in buckets.values():
            bucket.save()
```

**Comment by brlogan** [comment]
There ya go! That's pretty much what I was envisioning.
Nice use of `_get_collection()`.

**Comment by amateur-professional** [comment]
Decided to revisit this issue and try implementing some of your code.... aaaaand I wound up getting a couple different errors...

Using @dsnellgrove 's solution gives me the following traceback: 

```
Traceback (most recent call last):
  File "manage.py", line 10, in <module>
    execute_from_command_line(sys.argv)
  File "/usr/local/lib/python2.7/dist-packages/django/core/management/__init__.py", line 354, in execute_from_command_line
    utility.execute()
  File "/usr/local/lib/python2.7/dist-packages/django/core/management/__init__.py", line 346, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/usr/local/lib/python2.7/dist-packages/django/core/management/base.py", line 394, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/usr/local/lib/python2.7/dist-packages/django/core/management/base.py", line 445, in execute
    output = self.handle(*args, **options)
  File "/data/crits/crits/core/management/commands/bucket_counts.py", line 26, in handle
    for bucket_count in bucket_counts['result']:
TypeError: 'CommandCursor' object has no attribute '__getitem__'
```

I was under the impression the issue was with klass._get_collection() so I compensated by using @brlogan 's solution and received the following error:

```
Traceback (most recent call last):
  File "manage.py", line 10, in <module>
    execute_from_command_line(sys.argv)
  File "/usr/local/lib/python2.7/dist-packages/django/core/management/__init__.py", line 354, in execute_from_command_line
    utility.execute()
  File "/usr/local/lib/python2.7/dist-packages/django/core/management/__init__.py", line 346, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/usr/local/lib/python2.7/dist-packages/django/core/management/base.py", line 394, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/usr/local/lib/python2.7/dist-packages/django/core/management/base.py", line 445, in execute
    output = self.handle(*args, **options)
  File "/data/crits/crits/core/management/commands/bucket_counts.py", line 30, in handle
    bucket_counts = klass.objects().aggregate(pipe)
  File "/usr/local/lib/python2.7/dist-packages/mongoengine/queryset/base.py", line 1043, in aggregate
    return self._collection.aggregate(pipeline, cursor={}, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/pymongo/collection.py", line 1673, in aggregate
    result = self._command(sock_info, cmd, slave_ok)
  File "/usr/local/lib/python2.7/dist-packages/pymongo/collection.py", line 205, in _command
    read_concern=read_concern)
  File "/usr/local/lib/python2.7/dist-packages/pymongo/pool.py", line 213, in command
    read_concern)
  File "/usr/local/lib/python2.7/dist-packages/pymongo/network.py", line 99, in command
    helpers._check_command_response(response_doc, None, allowable_errors)
  File "/usr/local/lib/python2.7/dist-packages/pymongo/helpers.py", line 196, in _check_command_response
    raise OperationFailure(msg % errmsg, code, response)
pymongo.errors.OperationFailure: exception: pipeline element 0 is not an object
```
I don't really want to mess with it further after that since I already have a working solution ;)

I have to assume that both of your solutions - however useful they may be - are dependent on mongoengine/pymongo versions, whereas mine is more version agnostic. 

I was curious though about execution time; on our current dataset (over 3 million objects) this script executes in about 20 seconds. Still orders of magnitude better than the existing script!!

**Comment by brlogan** [comment]
Wow, I had forgotten about this. Could have saved me some re-work. Haha.
I think this is closed by #1046.

**Comment by amateur-professional** [comment]
After reading through your merge, yeah, I'm satisfied this is closed. Too bad I don't work at an org that uses CRITs anymore! This was a fun project while I did though :)
